cirq/ops/common_gates.py

```diff
@@ -322,6 +334,43 @@ def _act_on_(self, args: Any):
                                                         tableau.xs[:, q].copy())
             return True
 
+        if isinstance(args, clifford.ActOnStabilizerCHFormArgs):
+            if protocols.is_parameterized(self) or self.exponent % 0.5 != 0:
+                return NotImplemented
+            effective_exponent = self._exponent % 2
+            state = args.state
+            if effective_exponent == 0.5:
+                if ZPowGate()._act_on_(args) is True:
+                    if H._act_on_(args) is True:
+                        state.omega *= (1 + 1j) / (2**0.5)  # type: ignore
+                        return True
+                    # Only one of the delegated _act_on_ succeeded, so the
+                    # state might have been corrupted, return False instead of
+                    # NotImplemented.
+                    return False  # coverage: ignore
```

Comment:
Make a test that triggers it, or assert it doesn't happen.

Reply:
Assert-ed!

I was creating a tri-state where `False` denotes a corruption of state or an error and results in https://github.com/quantumlib/Cirq/blob/master/cirq/protocols/act_on_protocol.py#L102

NotImplemented allows the protocol to try the fallback strategy.

If the first `act_on` fails, the state has not been modified and we are free to try the fallback. If the first `act_on` succeeds but some subsequent one fails, the state might be unintentionally modified and we should fail fast.

But based on comments by @balopat and you, this tri-state behavior might be too messy while adding minimal value. 

I decided to do away with it and `assert` in the way you suggest here. This will cause us to fail when any of the listed gates fails to `act_on`(including the first one). But given that these are well-established Clifford gates that should have `_act_on_` implemented, even if the first one fails, it is actually more prudent to assert fail right here rather than keeping an option of a fallback strategy.